File: backend/app/schemas/data_quality.py
"""Pydantic schemas for data quality / missing data diagnostics."""

from pydantic import BaseModel, Field


# ── Request schemas ────────────────────────────────────────────────────────────


class DataQualityRequest(BaseModel):
    column_ids: list[int]
    include_na_as_missing: bool = True
    include_empty_as_missing: bool = True


class MissingPatternsRequest(DataQualityRequest):
    max_patterns: int = Field(default=50, ge=1, le=200)


# ── Response schemas ───────────────────────────────────────────────────────────


class VariableMissingSummary(BaseModel):
    column_id: int
    variable_name: str
    full_label: str
    dataset_id: int
    dataset_name: str
    column_type: str
    n_total: int
    n_valid: int
    n_missing: int
    pct_missing: float
    n_empty: int
    n_na: int


class MissingSummaryResponse(BaseModel):
    variables: list[VariableMissingSummary]
    total_rows: int
    total_cells: int
    total_missing: int
    overall_pct_missing: float


class PatternRow(BaseModel):
    pattern: list[bool]
    count: int
    pct: float


class MissingPatternsResponse(BaseModel):
    column_ids: list[int]
    column_labels: list[str]
    patterns: list[PatternRow]
    total_rows: int
    n_unique_patterns: int
    truncated: bool


class McarEligibility(BaseModel):
    eligible: bool
    reason: str | None = None
    warning: str | None = None


class McarTestResult(BaseModel):
    chi2: float
    df: int
    p: float
    n: int
    n_patterns: int
    n_variables: int
    apa_string: str
    interpretation: str


class McarTestResponse(BaseModel):
    eligibility: McarEligibility
    result: McarTestResult | None = None

File: frontend/src/lib/api/data-quality.ts
import api from './client'

// Data Quality types
export interface VariableMissingSummary {
  column_id: number
  variable_name: string
  full_label: string
  dataset_id: number
  dataset_name: string
  column_type: string
  n_total: number
  n_valid: number
  n_missing: number
  pct_missing: number
  n_empty: number
  n_na: number
}

export interface MissingSummaryResponse {
  variables: VariableMissingSummary[]
  total_rows: number
  total_cells: number
  total_missing: number
  overall_pct_missing: number
}

export interface PatternRow {
  pattern: boolean[]
  count: number
  pct: number
}

export interface MissingPatternsResponse {
  column_ids: number[]
  column_labels: string[]
  patterns: PatternRow[]
  total_rows: number
  n_unique_patterns: number
  truncated: boolean
}

export interface McarEligibility {
  eligible: boolean
  reason: string | null
  warning: string | null
}

export interface McarTestResult {
  chi2: number
  df: number
  p: number
  n: number
  n_patterns: number
  n_variables: number
  apa_string: string
  interpretation: string
}

export interface McarTestResponse {
  eligibility: McarEligibility
  result: McarTestResult | null
}

export interface DataQualityRequestBody {
  column_ids: number[]
  include_na_as_missing: boolean
  include_empty_as_missing: boolean
}

// API functions - Data Quality
export const dataQualityApi = {
  summary: (projectId: number, data: DataQualityRequestBody) =>
    api.post<MissingSummaryResponse>(`/projects/${projectId}/data-quality/summary`, data).then(r => r.data),

  patterns: (projectId: number, data: DataQualityRequestBody & { max_patterns?: number }) =>
    api.post<MissingPatternsResponse>(`/projects/${projectId}/data-quality/patterns`, data).then(r => r.data),

  mcarTest: (projectId: number, data: DataQualityRequestBody) =>
    api.post<McarTestResponse>(`/projects/${projectId}/data-quality/mcar-test`, data).then(r => r.data),

  summaryCsv: (projectId: number, params: {
    column_ids: number[]
    include_na_as_missing: boolean
    include_empty_as_missing: boolean
  }) =>
    api.get(`/projects/${projectId}/data-quality/summary/csv`, {
      params: {
        column_ids: params.column_ids.join(','),
        include_na_as_missing: params.include_na_as_missing,
        include_empty_as_missing: params.include_empty_as_missing,
      },
      responseType: 'blob',
    }).then(r => r.data as Blob),
}
